src/drivers/tty/termios_ops.c

```diff
@@ -135,33 +151,76 @@ int termios_handle_erase(const struct termios *t, char ch, int *raw_mode,
 
 			i_ring->head -= erase_len - buflen;
 			ring_fixup_head(i_ring, buflen);
+
+			while (erase_len--) {
+				result += termios_echo_erase(t, o_ring, o_buff, buflen);
+			}
 		}
 
-		return erase_len;
+		return result;
 	}
 
 	return -1;
 }
 
-int termios_input_status(const struct termios *t, char ch,
-	struct ring *i_ring, struct ring *i_canon_ring) {
+int termios_input_status(const struct termios *t,
+	struct termios_i_buff *b, char ch) {
 	
 	int raw_mode = !TIO_L(t, ICANON);
 	int is_eol = (ch == '\n' || ch == t->c_cc[VEOL]);
 	int got_data = (raw_mode || is_eol || ch == t->c_cc[VEOF]);
 
 	if (got_data) {
-		i_canon_ring->head = i_ring->head;
+		b->canon_ring->head = b->ring->head;
 
 		if (raw_mode) {
 			/* maintain it empty */
-			i_canon_ring->tail = i_canon_ring->head;
+			b->canon_ring->tail = b->canon_ring->head;
 		}
 	}
 
 	return got_data;
 }
 
+int termios_input(const struct termios *t, char ch, int *is_ready,
```

Comment:
`termios_input` now returns bool `got_data`. You could pack `is_ready` and `got_data` in one int result, calculated from ORing values like 
```
#define TIO_RES_GOT_DATA (1 << 0)
#define TIO_RES_GOT_ECHO (1 << 1)
```